### ISSUE TITLE
Is signing possible through GitHub Actions?

### ISSUE THREAD

**Issue by n8henrie** [quicksilver/Quicksilver#2576]
Since GitHub actions seems readily able to compile and provide a universal binary using the `qsrelease` script, the next logical step would be to see if we can use something like a repository secret to allow signing (and hopefully avoid all the issues about being unable to open since Big Sur).

<https://docs.github.com/en/actions/security-guides/encrypted-secrets>

- [x] Is there a way to keep a signing identity in a GitHub secret?
- [x] Discussion: Is this desirable? Would allow a "rogue contributor" to create a signed copy by pushing a new tag
    - In practice, this doesn't seem much different than not signing them at all, which is what's happening now

**Comment by gliptak** [comment]
https://localazy.com/blog/how-to-automatically-sign-macos-apps-using-github-actions

**Comment by skurfer** [comment]
I didn’t want to create a second issue related to signing, but I wanted to document what we know and have tried (since most of it is probably only in my head) so we can start talking about solutions. Maybe we should rename this issue to “Code Signing Status” and move this to the description, but here goes…

Releases were signed for a long time after Gatekeeper was first introduced. Code signing stopped working somewhere around Xcode 10. For a while, we got away with building QS in a VM running 10.13 with Xcode 9, but even that stopped working. (The signing, not the builds.)

Most people have probably always used Xcode to sign their apps. That will not work for us, so we have always told Xcode not to sign anything, then done it on the command-line in the `qsrelease` script. Why?

  1. Xcode requires you to define the signing cert, but this project is worked on by several independent devs and only one of them (me) has the cert we use.
  2. This project is configured to build several things then put them together after the fact. Xcode will sign `Quicksilver.app` as soon as it’s built, then copy something (like the frameworks or the Core Support plug-in) into the application bundle, rendering the signature invalid. Is the way this project does things “wrong” or unusual? I have no idea, but it must be because Apple clearly assumes a different setup.

The first reason surely has a solution because there must be a ton of projects like this.

On the second reason, we’re pretty screwed right now. As far as I can tell there is no way to tell a recent version of Xcode not to do any signing at all, so the command-line method will simply complain that the app is already signed. (But the signature won’t be valid because of all the other things the build process does.) This is why I say we should be focusing our efforts on redesigning the Xcode project. Maybe if we _could_ let Xcode do the signing, it would actually work.

Even if we could still do it manually, like I said, the command line technique just stopped giving us valid signatures for no apparent reason. Here’s some detail on that.

For years, this worked:

```console
$ codesign -s "Developer ID Application" --deep Quicksilver.app
```

But now, you’ll consistently see this:

```console
$ spctl -av Quicksilver.app
Quicksilver.app: rejected
```

One recommendation (and the official advice) is to sign the app “inside out”, which is like this:

```console
$ codesign -s "Developer ID Application" Quicksilver.app/Contents/Frameworks/*.framework
$ codesign -s "Developer ID Application" Quicksilver.app/Contents/PlugIns/*.qsplugin
$ codesign -s "Developer ID Application" Quicksilver.app/Contents/Resources/QSDroplet.app
$ codesign -s "Developer ID Application" Quicksilver.app/Contents/Library/Automator/*.action
$ codesign -s "Developer ID Application" Quicksilver.app/Contents/Resources/*.plist
$ codesign -s "Developer ID Application" Quicksilver.app/Contents/Info.plist
$ codesign -s "Developer ID Application” Quicksilver.app
```

If you try that process and verify the results after each step, everything reports valid as you go, and the end result for the whole app looks good according to `codesign`.

```console
$ codesign --verify --deep --strict --verbose=4 Quicksilver.app
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Primer Interface.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Finder Plugin.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Keyboard Triggers.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Bezel Interface.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Keyboard Triggers.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Core Support.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Primer Interface.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Bezel Interface.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Finder Plugin.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Core Support.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Get Quicksilver Selection.action
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Send to Quicksilver.action
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSEffects.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Get Quicksilver Selection.action
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Send to Quicksilver.action
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSEffects.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSFoundation.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSCore.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSCore.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSInterface.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSFoundation.framework/Versions/Current/Frameworks/FMDB.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSInterface.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSFoundation.framework/Versions/Current/Frameworks/FMDB.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSFoundation.framework/Versions/Current/.
Quicksilver.app: valid on disk
Quicksilver.app: satisfies its Designated Requirement
```

But on the same build…

```console
$ spctl -a -vv Quicksilver.app
Quicksilver.app: rejected
origin=Developer ID Application: Rob McBroom (ATVCND8EAP)
```

For what it’s worth, `spctl` accepts the DMG which I’m signing with the same cert. I’ve also checked the app with Patrick Wardle’s [What’s Your Sign?][1] and Apple’s [`check-signature` tool][2]. They both agree with `codesign` that the app is valid, but the `spctl` check fails and if you try to open it, Gatekeeper will tell you it’s “from an unknown developer”.

[1]: https://objective-see.com/products/whatsyoursign.html
[2]: https://developer.apple.com/library/archive/technotes/tn2206/_index.html#//apple_ref/doc/uid/DTS40007919-CH1-TNTAG211

**Comment by gliptak** [comment]
@skurfer thank you for capturing the complexities of the signing process.

Below two links might have pointers on how skip `Xcode` signing, possibly allowing for better commandline `codesign` results.

https://stackoverflow.com/questions/52602009/xcode-10-where-is-the-dont-code-sign-option-in-code-signing-identity
https://stackoverflow.com/questions/31039513/how-can-i-skip-code-signing-for-development-builds-in-xcode/31985749

**Comment by skurfer** [comment]
Thanks @gliptak. The advice in the first link appears to still be valid in Xcode 13.1. Unfortunately, the manually signed app still isn’t accepted by `spctl`.

**Comment by n8henrie** [comment]
I've been mucking around all morning with this in https://github.com/quicksilver/Quicksilver/pull/2582

Using a self-signed cert, I'm in the same position. Codesign is quite happy, spctl is not:

```console
$ codesign --verify --verbose --strict Quicksilver.app/
Quicksilver.app/: valid on disk
Quicksilver.app/: satisfies its Designated Requirement
$ spctl --assess -vvvvvvvv --type open --context context:primary-signature ./Quicksilver.app/
./Quicksilver.app/: rejected
origin=Quicksilver Test
```

Breadcrumbs for myself:
- I added my self-signed cert to spctl:
```
$ export SIGNING_IDENTITY="Quicksilver Test"
$ spctl --add --requirement "certificate leaf[subject.CN] = \"${SIGNING_IDENTITY}\"" --label "QuicksilverTest"
```
- Looks to have worked:
```
$ spctl --list | grep -i quicksilver
2728[QuicksilverTest] P0 allow execute
        certificate leaf[subject.CN] = "Quicksilver Test"
```
- spctl fails all the same
- Is `allow execute` an issue here? Don't I need `--type open` for a dmg?
- Can remove this with `spctl --remove --label QuicksilverTest`

**Comment by gliptak** [comment]
@n8henrie While I'm not sure how relevant below is, I noticed the following while reviewing the build settings:

https://github.com/gliptak/Quicksilver/runs/4184969754?check_suite_focus=true#step:3:11

`AD_HOC_CODE_SIGNING_ALLOWED = YES`

Should this (and maybe other signing related settings like `XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY` and `XCODE_ATTRIBUTE_CODE_SIGNING_ALLOWED`) be overridden to prevent Xcode from signing?

**Comment by n8henrie** [comment]
Just thinking about de307833f7fb25a556e0c2733f0d83006e51e933 in the setting of this issue -- is there a way that a malicious PR could expose a secret? I'm sure this is well covered somewhere, just want to make sure we don't embed a private signing key as a repo secret and then automatically run a PR for:

```bash
echo "${env.MY_PRIVATE_KEY}"
```

**Comment by gliptak** [comment]
You probably already reviewed the official writeup on secrets https://docs.github.com/en/actions/security-guides/encrypted-secrets
Operationally, maybe PR artifacts won't get signed (or signed by "fake" developer credentials) while release artifacts are signed.

**Comment by n8henrie** [comment]
Instructions from an official source: <https://docs.github.com/en/actions/deployment/deploying-xcode-applications/installing-an-apple-certificate-on-macos-runners-for-xcode-development>, looks very similar to the above.

Testing at <https://github.com/n8henrie/Quicksilver/actions/runs/1485234862>

**Comment by gliptak** [comment]
Reading more
"Note: With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is triggered from a forked repository."
so PR work flow cannot expose encrypted secrets

**Comment by n8henrie** [comment]
Cool, got it working with my self-signed cert: <https://github.com/n8henrie/Quicksilver/runs/4275197708?check_suite_focus=true>

> + codesign --verify --deep --strict --verbose=1 Quicksilver.app
Quicksilver.app: valid on disk
Quicksilver.app: satisfies its Designated Requirement
+ spctl --assess --verbose --type open --type exec ./Quicksilver.app
./Quicksilver.app: rejected

`spctl ... Quicksilver.app` likely failing because I haven't run `spctl --add <my self-signed cert>` on the runner, but working locally (so I *believe* it should work fine with a valid Apple-approved cert).

@skurfer -- if you're interested in adding the cert to CI to automatically sign, the steps are well explained in the links above. The secrets I used in the branch above were `KEYCHAIN_PASSWORD`, `MACOS_CERTIFICATE`, `MACOS_CERTIFICATE_PASSWORD`. Other than a typo, my other bump in testing was that I initially exported only the private key; have to make sure to also select the certificate (both get exported into the single file).

I'll wait until `cleanup_qsrelease` is merged before making a PR. Like you suggested, I'll start just pushing branches to `Quicksilver` going forward, I'm tripping over my own feet with the `n8henrie/Quicksilver` dance.

**Comment by n8henrie** [comment]
> Yeah, I’m happy to add the real cert to see what happens.

Sounds good. Once https://github.com/quicksilver/Quicksilver/pull/2582 and https://github.com/quicksilver/Quicksilver/pull/2577 are merged I'll packed this up into another PR and can point you towards the next steps in the above links.

> Sorry I haven’t been around as mush as I’d like.

No worries, you've been great.

**Comment by skurfer** [comment]
I’ve added the 3 secrets to the Quicksilver organization. Let me know if I need to adjust anything.

**Comment by n8henrie** [comment]
Sweet! I just pushed a new branch: https://github.com/quicksilver/Quicksilver/tree/signing

It's based on https://github.com/quicksilver/Quicksilver/pull/2582, so if it gets merged into master should also close that PR / issue.

**Comment by n8henrie** [comment]
Apparently new branches don't trigger CI ¯\\_(ツ)_/¯ so I made a new PR to trigger a build: https://github.com/quicksilver/Quicksilver/pull/2587

Had previously set CI to use a default `SIGNING_IDENTITY` of `-` (ad-hoc signing), removed that as well.

**Comment by n8henrie** [comment]
Well that's cute: [`No space left on device`](https://github.com/quicksilver/Quicksilver/runs/4463036172?check_suite_focus=true)

At least we've found a way to reproduce that.

**Comment by n8henrie** [comment]
I merged the updated master back into signing and re-removed the ad-hoc signing identity that got re-added somewhere.

Also changed the `|| true`s to `|| echo $?`s to hopefully give a better idea of what parts of the signing process are working; left on debug mode for the moment.

**Comment by n8henrie** [comment]
hmm -- spctl rejects with `source=Unnotarized Developer ID`: <https://github.com/quicksilver/Quicksilver/runs/4619882260?check_suite_focus=true#step:3:81>

However, I'm able to open the resulting DMG: <https://github.com/quicksilver/Quicksilver/suites/4748416340/artifacts/131411427>

I get a warning on first launch (have to right-click -> open), but at least the DMG opens.

I wonder what does `Unnotarized Developer ID` mean?

**Comment by n8henrie** [comment]
I've been seeing a few things about "stapling" with e.g. `xcrun stapler staple Quicksilver.app` after Apple "notarized" a DMG. Perhaps this is something we need to do to avoid all the warning dialogs?

<https://stackoverflow.com/a/62085946/1588795>

**Comment by pjrobertson** [comment]
Hopefully Rob can take a look, I don't have an Apple Developer Signing Identity.

**Comment by n8henrie** [comment]
Creating a self-signed cert: https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Procedures/Procedures.html

from above:

```console
$ export SIGNING_IDENTITY="Quicksilver Test"
$ spctl --add --requirement "certificate leaf[subject.CN] = \"${SIGNING_IDENTITY}\"" --label "QuicksilverTest"
```

Not working yet.

**Comment by n8henrie** [comment]
A little tinkering: `execute` seems to work for `.app` (search for `+ spctl --assess --type execute --context`)

```bash
#!/usr/bin/env bash

set -x

main() {
  local target command commands
  targets=("/tmp/QS/build/Release/Quicksilver.app" "/tmp/QS/build/Release/Quicksilver 1.6.1.dmg")

  export SIGNING_IDENTITY="Quicksilver Test"

  for target in "${targets[@]}"; do
    codesign --force --deep --sign "${SIGNING_IDENTITY}" -vvvv "${target}"
    codesign --verify -vvvv "${target}"
    commands=(install execute open)
    for cmd in "${commands[@]}"; do
      spctl --assess --type "${cmd}" --context context:primary-signature -vvvv "${target}"
    done
  done
}
main "$@"
```

```
$ ./spctl.sh
+ main
+ local target command commands
+ targets=("/tmp/QS/build/Release/Quicksilver.app" "/tmp/QS/build/Release/Quicksilver 1.6.1.dmg")
+ export 'SIGNING_IDENTITY=Quicksilver Test'
+ SIGNING_IDENTITY='Quicksilver Test'
+ for target in "${targets[@]}"
+ codesign --force --deep --sign 'Quicksilver Test' -vvvv /tmp/QS/build/Release/Quicksilver.app
/tmp/QS/build/Release/Quicksilver.app: replacing existing signature
/tmp/QS/build/Release/Quicksilver.app: signed app bundle with Mach-O universal (x86_64 arm64) [com.blacktree.Quicksilver]
+ codesign --verify -vvvv /tmp/QS/build/Release/Quicksilver.app
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Keyboard Triggers.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Keyboard Triggers.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Primer Interface.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Primer Interface.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Finder Plugin.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Finder Plugin.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Bezel Interface.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Bezel Interface.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Core Support.qsplugin
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/PlugIns/Core Support.qsplugin
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Send to Quicksilver.action
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Send to Quicksilver.action
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Get Quicksilver Selection.action
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Library/Automator/Get Quicksilver Selection.action
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSEffects.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSEffects.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSFoundation.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSFoundation.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCTAutomationSupport.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCTAutomationSupport.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCUIAutomation.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCUIAutomation.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCUnit.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCUnit.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSCore.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSCore.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSInterface.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/QSInterface.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCTestCore.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCTestCore.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCTest.framework/Versions/Current/.
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/XCTest.framework/Versions/Current/.
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/libXCTestSwiftSupport.dylib
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/libXCTestSwiftSupport.dylib
--prepared:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/libXCTestBundleInject.dylib
--validated:/private/tmp/QS/build/Release/Quicksilver.app/Contents/Frameworks/libXCTestBundleInject.dylib
/tmp/QS/build/Release/Quicksilver.app: valid on disk
/tmp/QS/build/Release/Quicksilver.app: satisfies its Designated Requirement
+ commands=(install execute open)
+ for cmd in "${commands[@]}"
+ spctl --assess --type install --context context:primary-signature -vvvv /tmp/QS/build/Release/Quicksilver.app
/tmp/QS/build/Release/Quicksilver.app: rejected
origin=Quicksilver Test
+ for cmd in "${commands[@]}"
+ spctl --assess --type execute --context context:primary-signature -vvvv /tmp/QS/build/Release/Quicksilver.app
/tmp/QS/build/Release/Quicksilver.app: accepted
source=QuicksilverTest
origin=Quicksilver Test
+ for cmd in "${commands[@]}"
+ spctl --assess --type open --context context:primary-signature -vvvv /tmp/QS/build/Release/Quicksilver.app
/tmp/QS/build/Release/Quicksilver.app: rejected
origin=Quicksilver Test
+ for target in "${targets[@]}"
+ codesign --force --deep --sign 'Quicksilver Test' -vvvv '/tmp/QS/build/Release/Quicksilver 1.6.1.dmg'
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: replacing existing signature
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: signed disk image [Quicksilver 1]
+ codesign --verify -vvvv '/tmp/QS/build/Release/Quicksilver 1.6.1.dmg'
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: valid on disk
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: satisfies its Designated Requirement
+ commands=(install execute open)
+ for cmd in "${commands[@]}"
+ spctl --assess --type install --context context:primary-signature -vvvv '/tmp/QS/build/Release/Quicksilver 1.6.1.dmg'
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: rejected
origin=Quicksilver Test
+ for cmd in "${commands[@]}"
+ spctl --assess --type execute --context context:primary-signature -vvvv '/tmp/QS/build/Release/Quicksilver 1.6.1.dmg'
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: rejected (the code is valid but does not seem to be an app)
origin=Quicksilver Test
+ for cmd in "${commands[@]}"
+ spctl --assess --type open --context context:primary-signature -vvvv '/tmp/QS/build/Release/Quicksilver 1.6.1.dmg'
/tmp/QS/build/Release/Quicksilver 1.6.1.dmg: rejected
```

**Comment by pjrobertson** [comment]
Marking this as fixed by #2655 :)
